### PR TITLE
Fix bugs in parsing `with header row` and `partitioned by`

### LINKED ISSUE #4262: `create external table` should fail to parse if syntax is incorrect
**Describe the bug**
When using `create external table` with csv files we can specify `with header row` to indicate that the file contains headers.

I ran this and forgot the syntax and omitted the word `row` and it did not process the header row and also did not fail and tell me that the syntax was incorrect.

**To Reproduce**

```
❯ create external table taxi_zone stored as csv with header location "/mnt/bigdata/nyctaxi/taxi+_zone_lookup.csv";
0 rows in set. Query took 0.008 seconds.
❯ select * from taxi_zone limit 10;
+------------+---------------+-------------------------+--------------+
| column_1   | column_2      | column_3                | column_4     |
+------------+---------------+-------------------------+--------------+
| LocationID | Borough       | Zone                    | service_zone |
| 1          | EWR           | Newark Airport          | EWR          |
| 2          | Queens        | Jamaica Bay             | Boro Zone    |
| 3          | Bronx         | Allerton/Pelham Gardens | Boro Zone    |
| 4          | Manhattan     | Alphabet City           | Yellow Zone  |
| 5          | Staten Island | Arden Heights           | Boro Zone    |
| 6          | Staten Island | Arrochar/Fort Wadsworth | Boro Zone    |
| 7          | Queens        | Astoria                 | Boro Zone    |
| 8          | Queens        | Astoria Park            | Boro Zone    |
| 9          | Queens        | Auburndale              | Boro Zone    |
+------------+---------------+-------------------------+--------------+
```

**Expected behavior**
SQL should have failed to parse.

**Additional context**
None

**Task list**
- [x] with header row (#4268)
- [ ] compression type
- [x] partitioned by (#4268)

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>

# Which issue does this PR close?

Part of #4262.

# Rationale for this change

# What changes are included in this PR?

Fix the errors in parsing `with header row` and `partitioned by`.
# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by HaoYang670** [comment]
For `COMPRESSION TYPE`, things go a little different, as `COMPRESSION` isn't a `Keyword` in sqlparser. We can either wait for the `sqlparser` to update, or we can fix in datafusion with a different way:
```rust
   fn parse_has_file_compression_type(&mut self) -> Result<bool, ParserError> {
        // return error if finding `compression` but not find `TYPE`
    }
```

File: datafusion/sql/src/parser.rs [hunk 1]
```diff
@@ -715,6 +715,17 @@ mod tests {
             "CREATE EXTERNAL TABLE t STORED AS x OPTIONS ('k1' 'v1', k2 v2, k3) LOCATION 'blahblah'";
         expect_parse_error(sql, "sql parser error: Expected literal string, found: )");
 
+        // Error case: `with header` is an invalid syntax
+        let sql = "CREATE EXTERNAL TABLE t STORED AS CSV WITH HEADER LOCATION 'abc'";
+        expect_parse_error(sql, "sql parser error: Expected LOCATION, found: WITH");
+
+        // Error case: `with header` is an invalid syntax
```
Comment (andygrove) [inline datafusion/sql/src/parser.rs:722]: nit: comment seems incorrect here

**Review by alamb — APPROVED** [review]
Thanks @HaoYang670

**Comment by alamb** [comment]
CI failure due to https://github.com/apache/arrow-datafusion/issues/4294 (not this PR)

**Pull request merged**